### PR TITLE
[JIT] Introduce a fake Tensor creation node for IR unit tests

### LINKED ISSUE #33500: Use fake nodes in tests that hard code IR
A number of JIT C++ tests hard-code IR to test aliasing and dce, e.g.,

```
      script::parseIR(
          R"IR(
  graph():
    %10 : bool = prim::Constant[value=0]()
    %8 : Device? = prim::Constant()
    %4 : int? = prim::Constant()
    %0 : int = prim::Constant[value=2]()
    %1 : int = prim::Constant[value=3]()
    %2 : int[] = prim::ListConstruct(%0, %1)
    %x : Tensor = aten::rand(%2, %4, %4, %8, %10)
    %12 : int[] = prim::ListConstruct(%0, %1)
    %y : Tensor = aten::rand(%12, %4, %4, %8, %10)
    %22 : int[] = prim::ListConstruct(%0, %1)
    %z : Tensor = aten::rand(%22, %4, %4, %8, %10)
    %32 : int[] = prim::ListConstruct(%0, %1)
    %fresh : Tensor = aten::rand(%32, %4, %4, %8, %10)
    %foo : Tensor[] = prim::ListConstruct(%x, %y)
    %43 : Tensor[] = aten::append(%foo, %z)
    return ()
  )IR",
```

in `test/cpp/jit/test_alias_analysis.cpp`

IR represented at this stage has had all default arguments resolved into explicit arguments in the IR. Unfortunately, this is a major source of test brittleness, because whenever the default arguments of a function change (even if the change is backwards compatible), every IR graph has to be updated (by hand!!) to reflect the default change. That's *really bad*, because it means any developer who wants to work on adding new keyword arguments to our factory functions needs to know about these JIT tests. Also, the error message when you get it wrong is crap:

```
RuntimeError: hasSpecialCase INTERNAL ASSERT FAILED at ../torch/csrc/jit/passes/alias_analysis.cpp:311,  please report a bug to PyTorch. We don't have an op for aten::ones but it isn't a special case. (analyzeImpl at ../torch/csrc/jit/passes/alias_analysis.cpp:311)
```

In the case of the tests here, we don't really need to use `aten::rand`, any factory function would do. So let's create a fake factory function and use that in the tests, instead of directly using real functions which may need to evolve.

There are multiple IR tests in this form in the alias analysis tests as well as the DCE tests.

cc @suo

### PR DESCRIPTION
**Pull request by SplitInfinity**
**Summary**
There is often a need to create a Tensor when writing IR by hand for JIT
optimisation pass unit tests. The only options for this today are real
Tensor creation functions like `aten::ones`. Any test that uses these functions
must also use the same default arguments as the Python/C++ API, which means
that all of the tests have to be updated when the API is updated. This commit
introduces a new primitive, `prim::MakeTestTensor` with schema `() -> Tensor` that
should be used in unit tests instead of real Tensor creation functions. This new
primitive has no public-facing API, so the maintenance burden is much lower.

**Testing**
This commit updates the alias analysis and DCE tests to use `prim::MakeTestTensor` instead of
`aten::rand`, `aten::ones`, and `aten::zeros`.

```
$ ./bin/test_jit
CUDA not available. Disabling CUDA and MultiCUDA tests
Note: Google Test filter = *-*_CUDA:*_MultiCUDA
[==========] Running 75 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 75 tests from JitTest
[ RUN      ] JitTest.ADFormulas
[       OK ] JitTest.ADFormulas (82 ms)
[ RUN      ] JitTest.Attributes
[       OK ] JitTest.Attributes (0 ms)
...
...
...
[ RUN      ] JitTest.LiteInterpreterPrim
[       OK ] JitTest.LiteInterpreterPrim (0 ms)
[ RUN      ] JitTest.LiteInterpreterLoadOrigJit
[       OK ] JitTest.LiteInterpreterLoadOrigJit (2 ms)
[----------] 75 tests from JitTest (150 ms total)

[----------] Global test environment tear-down
[==========] 75 tests from 1 test case ran. (150 ms total)
[  PASSED  ] 75 tests.
```

**Fixes**
This pull request fixes https://github.com/pytorch/pytorch/issues/33500.

### CODE REVIEW

**Comment by SplitInfinity** [comment]
Note that this is a repost of #33595. The PR was closed but ended up breaking the build because it did not merge nicely on top of #33851.

File: test/cpp/jit/test_base.cpp [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#include <test/cpp/jit/test_base.h>
+#include <test/cpp/jit/test_utils.h>
+
+#include "torch/csrc/jit/runtime/custom_operator.h"
```
Comment (SplitInfinity) [inline test/cpp/jit/test_base.cpp]: This is the only change.

**Comment by SplitInfinity** [comment]
cc: @suo @ezyang for code review. For some reason I can't add reviewers after publishing the PR.

File: test/cpp/jit/test_base.cpp [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  c10::OperatorOptions result;
+  result.setAliasAnalysis(c10::AliasAnalysisKind::FROM_SCHEMA);
+  return result;
+}
+
+RegisterOperators reg({
+    Operator(
+        "prim::MakeTestTensor() -> Tensor",
+        [](Stack& stack) {
+          push(stack, at::Tensor());
```
Comment (ezyang) [inline test/cpp/jit/test_base.cpp]: NB: if you ever try to actually run these tests these are going to produce undefined tensors, which will probably blow up in unexpected ways. So it's probably worth documenting MakeTestTensor to this effect.

**Review by ezyang — APPROVED** [review]
This seems good, but I'd beef up the documentation. Also probably good to get a real JIT reviewer's look at this.

**Pull request closed without merging**